widget/card.go

```diff
@@ -0,0 +1,204 @@
+// Layout the components of the card container.
+func (c *cardRenderer) Layout(size fyne.Size) {
+	pos := fyne.NewPos(theme.Padding()/2, theme.Padding()/2)
+	size = size.Subtract(fyne.NewSize(theme.Padding(), theme.Padding()))
+	c.LayoutShadow(size, pos)
+
+	if c.card.Image != nil {
+		c.card.Image.Move(pos)
+		c.card.Image.Resize(fyne.NewSize(size.Width, cardMediaHeight))
+		pos.Y += cardMediaHeight
+	}
+
+	size.Width -= theme.Padding() * 4
+	pos.X += theme.Padding() * 2
+	pos.Y += theme.Padding()
+
+	if c.card.Title != "" {
+		height := c.header.MinSize().Height
+		c.header.Move(pos)
+		c.header.Resize(fyne.NewSize(size.Width, height))
+		pos.Y += height + theme.Padding()
+	}
+
+	if c.card.SubTitle != "" {
+		height := c.subHeader.MinSize().Height
+		c.subHeader.Move(pos)
+		c.subHeader.Resize(fyne.NewSize(size.Width, height))
+		pos.Y += height + theme.Padding()
+	}
+
+	size.Width += theme.Padding() * 2
+	pos.X -= theme.Padding()
+	pos.Y += theme.Padding()
+
+	if c.card.Content != nil {
+		height := size.Height - pos.Y - theme.Padding()*3
+		c.card.Content.Move(pos)
+		c.card.Content.Resize(fyne.NewSize(size.Width, height))
+	}
+}
+
+// applyTheme updates this button to match the current theme
+func (c *cardRenderer) applyTheme() {
+	if c.header != nil {
+		c.header.TextSize = int(float32(theme.TextSize()) * 1.7)
+		c.header.Color = theme.TextColor()
+	}
+	if c.subHeader != nil {
+		c.subHeader.TextSize = theme.TextSize()
+		c.subHeader.Color = theme.TextColor()
+	}
+}
+
+func (c *cardRenderer) BackgroundColor() color.Color {
+	return theme.BackgroundColor()
+}
+
+func (c *cardRenderer) Refresh() {
+	c.header.Text = c.card.Title
+	c.header.Refresh()
+	c.subHeader.Text = c.card.SubTitle
+	c.subHeader.Refresh()
+
+	c.applyTheme()
+	c.Layout(c.card.Size())
+	canvas.Refresh(c.card.super())
+}
+
+// CardContainer widget groups title, subtitle with content and a header image
+type CardContainer struct {
+	BaseWidget
+	Title, SubTitle string
+	Image           *canvas.Image
+	Content         fyne.CanvasObject
+}
+
+// MinSize returns the size that this widget should not shrink below
+func (c *CardContainer) MinSize() fyne.Size {
+	c.ExtendBaseWidget(c)
+	return c.BaseWidget.MinSize()
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer
+func (c *CardContainer) CreateRenderer() fyne.WidgetRenderer {
+	c.ExtendBaseWidget(c)
+
+	header := canvas.NewText(c.Title, theme.TextColor())
+	header.TextStyle.Bold = true
+	subHeader := canvas.NewText(c.SubTitle, theme.TextColor())
+
+	objects := []fyne.CanvasObject{header, subHeader}
+	if c.Image != nil {
+		objects = append(objects, c.Image)
+	}
+	if c.Content != nil {
+		objects = append(objects, c.Content)
+	}
+	r := &cardRenderer{widget.NewShadowingRenderer(objects, widget.CardLevel),
+		header, subHeader, c}
+	r.applyTheme()
+	return r
+}
+
+// SetText updates the main title for this card.
+func (c *CardContainer) SetTitle(text string) {
+	c.Title = text
+
+	c.Refresh()
```

Comment:
It's not necessary to refresh the whole container, and depending on the contents this might be expensive... maybe the labels and image could be children of `CardContainer` rather than it's renderer.

Edit: the same applies to the other `Set*` methods below.

Reply:
Well, It's not that simple - if it were set to "" (or from "") then we do need a complete refresh.
Should that logic be in her to compare old and new values to do the different type of refreshes?

Reply:
Considering that the content could potentially be very expensive to refresh, I think it makes sense to avoid propagating it down the line needlessly. In particular, I imagine a common use case for cards would be as a container for a custom widget -- downstream developers may not be as careful about avoiding spurious refreshes, so I think for the best developer experience we should do as much of the legwork as possible in that area.